node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -1139,30 +1188,25 @@ async fn handle_collation_fetched_result(
 				"Failed to fetch collation.",
 			);
 
-			let (next_try, id) = if let Some(collations) = state.collations_per_relay_parent.get_mut(&relay_parent) {
-				if let Some(next_try) = collations.unfetched_collations.pop() {
-					next_try
-				} else if matches!(collations.status, CollationStatus::Fetching) {
-					collations.status = CollationStatus::Waiting;
-					return
-				} else {
-					tracing::error!(
-						target: LOG_TARGET,
-						status = ?collations.status,
-						"Expected status `CollationStatus::Fetching` but got unexpected status."
-					);
-					return
-				}
+			let fetch = if let Some(collations) = state.collations_per_relay_parent.get_mut(&relay_parent) {
+				collations.status.back_to_waiting();
+				collations.get_next_collation_to_fetch()
 			} else {
-				return
+				None
 			};
 
-			fetch_collation(ctx, state, next_try, id).await;
+			if let Some((next, id)) = fetch {
+				fetch_collation(ctx, state, next, id).await;
+			}
 
 			return
 		},
 	};
 
+	if let Some(collations) = state.collations_per_relay_parent.get_mut(&relay_parent) {
+		collations.status = CollationStatus::WaitingOnValidation;
```

Comment:
how can it be `None`?
can the status be `Seconded`?
I feel like we could the state transition function could use more type-safety and checked assumptions/invariants built-in

Reply:
It can not be none, but I also don't want to use an expect here.

> I feel like the state transition function could use more type-safety and checked assumptions/invariants built-in

Not sure what you mean here.

Reply:
> > I feel like the state transition function could use more type-safety and checked assumptions/invariants built-in

> Not sure what you mean here.

I don't have a suggestion how to refactor it yet, so nevermind.